core/server/data/migrations/versions/4.0/14-populate-members-paid-subscription-events-table.js

```diff
@@ -0,0 +1,112 @@
+const {createTransactionalMigration} = require('../../utils');
+
+module.exports = createTransactionalMigration(
+    async function up(knex) {
+        const allSubscriptions = await knex
+            .select(
+                'c.member_id',
+                's.status',
+                's.start_date',
+                's.updated_at',
+                's.created_at',
+                's.plan_id',
+                's.plan_amount',
+                's.plan_currency',
+                's.plan_interval'
+            )
+            .from('members_stripe_customers_subscriptions as s')
+            .join('members_stripe_customers as c', 'c.customer_id', '=', 's.customer_id');
+
+        function calculateMrrDelta({status, interval, amount}) {
+            if (status === 'trialing') {
+                return 0;
+            }
+            const modifier = status === 'canceled' ? -1 : 1;
+
+            if (interval === 'year') {
+                return modifier * (amount / 12);
+            }
+
+            if (interval === 'month') {
+                return modifier * amount;
+            }
+        }
+
+        const allEvents = allSubscriptions.reduce((allEventsAcc, subscription) => {
+            if (['incomplete', 'incomplete_expired'].includes(subscription.status)) {
+                return allEventsAcc;
+            }
+
+            const events = [];
+
+            if (subscription.status === 'trialing') {
+                const subscriptionCreatedEvent = {
+                    id: ObjectID.generate(),
+                    member_id: subscription.member_id,
+                    from_plan: null,
+                    to_plan: subscription.plan_id,
+                    currency: subscription.plan_currency,
+                    source: 'stripe',
+                    mrr_delta: calculateMrrDelta({
+                        status: subscription.status,
+                        amount: subscription.plan_amount,
+                        interval: subscription.plan_interval
+                    }),
+                    created_at: subscription.start_date
+                };
+                events.push(subscriptionCreatedEvent);
+            } else {
+                const subscriptionCreatedEvent = {
+                    id: ObjectID.generate(),
+                    member_id: subscription.member_id,
+                    from_plan: null,
+                    to_plan: subscription.plan_id,
+                    currency: subscription.plan_currency,
+                    source: 'stripe',
+                    mrr_delta: calculateMrrDelta({
+                        status: subscription.status,
+                        amount: subscription.plan_amount,
+                        interval: subscription.plan_interval
+                    }),
+                    created_at: subscription.start_date
+                };
+                events.push(subscriptionCreatedEvent);
+            }
+
+            if (subscription.status === 'canceled') {
```

Comment:
I don't think this will work as expected for cancelled events: if all of my subscriptions are cancelled, this migration will give me a negative aggregated MRR because we don't have the corresponding `created` events to cancel out. I think MRR delta for backfilling cancelled events should be 0.

Reply:
Okay maybe it's a bit too implicit, but the `else` block above will handle the initial subscription event so we would have an event with `+X` MRR and then the `-X` MRR afterwards

Reply:
I completely missed that - but won't we actually get two lots of `-X` MRR since the status is still cancelled?

Reply:
Yes you're totally right, I'll move the modifier out of the calculateMrrDelta function